File: 2022/python/day03.py
import sys
from functools import reduce
from pathlib import Path

from aoc_util import *


def split(rucksack):
    mid = len(rucksack) // 2
    return rucksack[:mid], rucksack[mid:]


def common_element(*rucksacks):
    sets = [set(rucksack) for rucksack in rucksacks]
    result = reduce(lambda a, b: a.intersection(b), sets)
    return result.pop()


def priority_value(letter):
    if letter.isupper():
        return 27 + ord(letter) - ord('A')
    else:
        return 1 + ord(letter) - ord('a')


def priority_sum(rucksacks):
    return sum(priority_value(common_element(*split(rucksack))) for rucksack in rucksacks)


def badge_sum(rucksacks):
    chunks = chunk(rucksacks, 3)
    return sum(priority_value(common_element(*c)) for c in chunks)


def main():
    input_path = Path(sys.argv[1])
    with open(input_path / "03.txt", "r") as f:
        print("Day 02:")
        rucksacks = f.read().strip().split("\n")
        a, b = priority_sum(rucksacks), badge_sum(rucksacks)
        print(f"\tPart 1: {a}")
        print(f"\tPart 2: {b}")


if __name__ == "__main__":
    main()

File: 2022/cpp/src/day03.cpp
#include <string>
#include <iostream>
#include <fstream>
#include <vector>
#include <numeric>
#include <set>

#include "day03.h"

int priority_value(char c) {
    if (std::isupper(c)) {
        return 27 + c - 'A';
    } else {
        return 1 + c - 'a';
    }
}

int rucksack_priority(std::string &rucksack) {
    auto mid = rucksack.begin() + std::distance(rucksack.begin(), rucksack.end()) / 2;
    std::set<char> c1{rucksack.begin(), mid};
    std::set<char> c2{mid, rucksack.end()};
    std::set<char> intersection;
    std::set_intersection(c1.begin(), c1.end(), c2.begin(), c2.end(),
                          std::inserter(intersection, intersection.begin()));
    return priority_value(*intersection.begin());
}

int rucksack_sum(const std::vector<std::string> &rucksacks) {
    std::vector<std::string> copy;
    std::copy(rucksacks.begin(), rucksacks.end(), std::back_inserter(copy));
    return std::transform_reduce(copy.begin(), copy.end(), 0,
                                 std::plus{}, rucksack_priority);
}

int group_priority(auto begin, auto end) {
    std::vector<std::set<char>> bags;
    std::for_each(begin, end, [&bags](auto it) {
        std::set<char> bag{it.begin(), it.end()};
        bags.push_back(bag);
    });

    auto result = std::reduce(bags.begin() + 1, bags.end(), *bags.begin(),
                              [](auto acc, auto bag) {
                                  std::set<char> next;
                                  std::set_intersection(bag.begin(), bag.end(), acc.begin(), acc.end(),
                                                        std::inserter(next, next.begin()));
                                  return next;
                              });
    return priority_value(*result.begin());
}

int badge_sum(const std::vector<std::string> &rucksacks) {
    int sum = 0;
    for (auto it = rucksacks.begin(); it < rucksacks.end(); it += 3) {
        sum += group_priority(it, it + 3);
    }
    return sum;
}

void day03_parse(std::istream &in, std::vector<std::string> &rucksacks) {
    for(std::string line; std::getline(in, line); ) {
        rucksacks.push_back(line);
    }
}

void day03(const std::string &filename) {
    std::ifstream in{filename};
    std::vector<std::string> rucksacks;
    day03_parse(in, rucksacks);
    auto a = rucksack_sum(rucksacks);
    auto b = badge_sum(rucksacks);
    std::cout << "Day 02:\n";
    std::cout << "\tPart 1: " << a << std::endl;
    std::cout << "\tPart 2: " << b << std::endl;
}
